### PR TITLE
Set edition 2018

### PR DESCRIPTION
**Pull request by bheylin**
I don't see what benefit not setting the `package.edition` provides. 
If it's not set then `edition = 2015` is assumed.

For now I'm going to set the edition to 2018 and fix up the issues that arise.

For the future, given how the [edition system works](https://doc.rust-lang.org/book/appendix-05-editions.html) I don't see any reason to not set the edition to 2021. 

Does anyone (@rikvdkleij, @maospr, @rudolphfroger, @ds-cbo, @remkop22) have any objections?

### CODE REVIEW

File: postgres-types/src/lib.rs [hunk 1]
```diff
@@ -282,10 +288,6 @@ mod uuid_08;
 #[cfg(feature = "with-uuid-1")]
 mod uuid_1;
 
-// The time::{date, time} macros produce compile errors if the crate package is renamed.
-#[cfg(feature = "with-time-0_2")]
```
Comment (bheylin) [inline postgres-types/src/lib.rs:286]: This in combination with the `cargo test --all-features` used by the previous version of the `github/worklows/ci.yml` would result in both `with-time-2_0` and `with-time-3_0` features being enable at the same time.
Which in turn would result in the `time v0.3` macros using `time v0.2` code.

It's best to test the time versions separately.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -103,4 +103,10 @@ jobs:
           key: test-target-${{ runner.os }}-${{ steps.rust-version.outputs.version }}-${{ hashFiles('Cargo.lock') }}y
       - run: cargo test --all
       - run: cargo test --manifest-path tokio-postgres/Cargo.toml --no-default-features
-      - run: cargo test --manifest-path tokio-postgres/Cargo.toml --all-features
+      # `with-time-0_2` and `with-time-0_3` features conflict with each other.
+      # the simplest approach is to run them separately
+      - run: cargo test --manifest-path tokio-postgres/Cargo.toml --no-default-features --features \
```
Comment (bheylin) [inline .github/workflows/ci.yml:108]: Testing both `with-time-2_0` and `with-time-3_0` features together is incorrect.
Given that the `time_2_0` dep is renamed to `time` and both features are enabled.
This means the `time` v3.x macros use the code from the `time` v2.x lib.

File: postgres-types/src/lib.rs (hunk 1 above)
Comment (rikvdkleij) [reply postgres-types/src/lib.rs:286]: I wonder why multiple version of `time`  and other libraries have to be supported. In my opinion it's sufficient to only support the latest version. I can imagine that when a new (breaking) version of libraries is released, temporarily 2 versions are supported.

Comment (bheylin) [reply postgres-types/src/lib.rs:286]: I guess it's something to do with the CVE that lingered for a while in `time` before a decent solution was found.
But now there's probably no point in supporting both versions.

TBH I get the impression that sfackler is not reviewing previous decisions that were made in `tokio-postgres`. 
He seems to be quite conservative in changing anything.

Comment (ds-cbo) [reply postgres-types/src/lib.rs:286]: Upstream wants to stay compatible for as low of an MSRV as possible. Time 0.3 bumps the MSRV compared to 0.2 so upstream would like to keep support for both, which imho is reasonable as long as it doesn't break on new compilers (like `eui48-0.4` did, for which I've removed the tests in upstream so it isn't necessary to run `cargo test --all-features`, but support is still there behind a feature gate for those who need it)

If anything, I would look into why 0.3 macros are used for 0.2 code. I would imagine that Rust has support for feature-gated macros as well and can be scoped appropriately. Simplifying the code base would work for our purposes but is very unlikely to be merged upstream, so we'd be stuck on our fork for the foreseeable future.

Taking one step back, what is the benefit of migrating everything to `edition = "2018"`? I thought the Rust compiler would happily keep compiling the 2015 edition without performance impacts, that's what editions are for after all.

Comment (bheylin) [reply postgres-types/src/lib.rs:286]: > Upstream wants to stay compatible for as low of an MSRV as possible. Time 0.3 bumps the MSRV compared to 0.2 so upstream would like to keep support for both, which imho is reasonable as long as it doesn't break on new compilers (like `eui48-0.4` did, for which I've removed the tests in upstream so it isn't necessary to run `cargo test --all-features`, but support is still there behind a feature gate for those who need it)

Ah that's why the two versions are used. That makes sense.

> If anything, I would look into why 0.3 macros are used for 0.2 code. I would imagine that Rust has support for feature-gated macros as well and can be scoped appropriately. Simplifying the code base would work for our purposes but is very unlikely to be merged upstream, so we'd be stuck on our fork for the foreseeable future.

The reason is that both features are enabled at the same time and the `extern crate` statement renames `time-2_0` to `time`. The `v0.3` macros then use whatever code is called `time`

> Taking one step back, what is the benefit of migrating everything to `edition = "2018"`? I thought the Rust compiler would happily keep compiling the 2015 edition without performance impacts, that's what editions are for after all.

A fair question, and that's the point of this PR as such, as ask that question and see what happens to the code when we 
'upgrade' to edition 2018.

I'm presuming that we will eventually branch off from upstream given that we're going to modify our fork to meet our needs, and ours needs are not sfacklers needs from what I see.

Upgrading edition is not necessary, but it does allow us to use the niceties that come along with more recent editions.
Things like as using `let-else` and better linting etc. 

These are certainly not reasons to break compatibility with upstream.
And we can always apply the results of the lints without actually upgrading to a more recent edition.

Comment (rikvdkleij) [reply postgres-types/src/lib.rs:286]: > Ah that's why the two versions are used. That makes sense.

The MSRV of time 0.3 is 1.48.  1.48 has been released in 19-11-2020. I just wonder what you would the maintenance burden for even older Rust versions.

Just my 2 cts :smile:

**Comment by bheylin** [comment]
I'm going to close this PR as it's not the right reason to break with upstream.
The macro issue is not actually an issue as `time v0.3` doesn't use macros!

**Pull request closed without merging**